Assets/Scripts/Map/Generators/Patrolling/VisibilityCache.cs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2025 MAEPS
+// 
+// This file is part of MAEPS
+// 
+// MAEPS is free software: you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+// 
+// MAEPS is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+// Public License for more details.
+// 
+// You should have received a copy of the GNU General Public License along
+// with MAEPS. If not, see http://www.gnu.org/licenses/.
+// 
+// Contributors: 
+// Henrik van Peet,
+// Mads Beyer Mogensen,
+// Puvikaran Santhirasegaram
+
+using System.Collections.Generic;
+
+using Maes.Utilities;
+
+using Unity.Collections;
+using Unity.Jobs;
+
+using UnityEngine;
+
+namespace Maes.Map.Generators.Patrolling
+{
+    public static class VisibilityCache
+    {
+        private static readonly Dictionary<(Bitmap, float), string> CachedVisibilities = new();
+
+        // Cache index to avoid overwriting the same file
+        private static int _cacheIndex = 0;
+
+        internal static Dictionary<Vector2Int, Bitmap> ComputeVisibilityCached(Bitmap map, float maxDistance = 0f)
+        {
+            Dictionary<Vector2Int, Bitmap> visibilities;
+            if (CachedVisibilities.TryGetValue((map, maxDistance), out var filepath))
+            {
+                // Load the cached visibility from the json file
+                return LoadFromCache(filepath);
+            }
+
+            visibilities = ComputeVisibility(map, maxDistance);
+            // Save the visibility to a json file
+            SaveToCache(map, maxDistance, visibilities);
+            return visibilities;
+        }
+
+        private static void SaveToCache(Bitmap map, float maxDistance, Dictionary<Vector2Int, Bitmap> visibilities)
+        {
+            if (_cacheIndex == 0)
+            {
+                // If the cache index is 0, we need to delete the old cache files
+                ClearCacheDirectory();
+            }
+
+            var json = JsonUtility.ToJson(visibilities);
+            var filename = $"visibility-{map.Width}X{map.Height}-{maxDistance}.json";
+            var path = System.IO.Path.Combine(GlobalSettings.MapCacheLocation, filename);
+
+            System.IO.Directory.CreateDirectory(GlobalSettings.MapCacheLocation);
+            System.IO.File.WriteAllText(path, json);
+            CachedVisibilities[(map, maxDistance)] = path;
```

Comment:
Hmm I dont think we should be storing anything in files. I think we can hold everything in memory. This way we wont have caching issues.

Reply:
But.. that's a bitmap stored in memory, for every non-wall tile, in the X*X size map, for every scenario that has a different map. Isn't that a bit much memory?

Reply:
we could probably just store the vertices

Reply:
If I make a 1000x1000 map of only non-wall tiles, this would be 1000 bitmaps of size 1000x1000. Even if they are BITmaps, that is 1000*1000*1000 bits =0,125 gigabytes

Reply:
Well what do we need the visibility calculations for? Just store the vertices and you are good.

Reply:
I suppose, but then you assume they are always generated the same way

Reply:
That is just as true for the visibility bitmaps though.

Reply:
You could probably just cache the result from GreedyMostVisibilityWaypointGenerator. That should be good for nearly all use cases. I don't even think we have any other waypoint generator.

Reply:
We currently don't, but some future group might add one ;)

Reply:
I think its better to just cache at that layer. We wont be adding any other waypoint generation algorithm.

Reply:
Even if we do we can add caching to that aswel

Reply:
Well handle it monday :)